src/grouping.jl

```diff
@@ -32,3 +32,43 @@ Base.getproperty(g::Grouping, prop::Symbol) = prop == :levels ? nothing : getfie
 function StatsModels.modelcols(::CategoricalTerm{Grouping}, d::NamedTuple)
     return error("can't create model columns directly from a Grouping term")
 end
+
+# copied and then adapted from StatsModels@463eb0a
+function StatsModels.ContrastsMatrix(
+    contrasts::Grouping, levels::AbstractVector{T}
+) where {T}
+
+    # if levels are defined on contrasts, use those, validating that they line up.
+    # what does that mean? either:
+    #
+    # 1. DataAPI.levels(contrasts) == levels (best case)
+    # 2. data levels missing from contrast: would generate empty/undefined rows.
+    #    better to filter data frame first
+    # 3. contrast levels missing from data: would have empty columns, generate a
+    #    rank-deficient model matrix.
+    c_levels = something(DataAPI.levels(contrasts), levels)
+
+    mismatched_levels = symdiff(c_levels, levels)
+    if !isempty(mismatched_levels)
+        throw(
+            ArgumentError(
+                "contrasts levels not found in data or vice-versa: " *
+                "$mismatched_levels." *
+                "\n  Data levels ($(eltype(levels))): $levels." *
+                "\n  Contrast levels ($(eltype(c_levels))): $c_levels",
+            ),
+        )
+    end
+
+    # do conversion AFTER checking for levels so users get a nice error message
+    # when they've made a mistake with the level types
+    c_levels = convert(Vector{T}, c_levels)
+
+    n = length(c_levels)
+    # not validating this allows for prediction of only a single level of the grouping factor
+
+    baseind = 1
+    tnames = StatsModels.termnames(contrasts, c_levels, baseind)
+    mat = StatsModels.contrasts_matrix(contrasts, baseind, n)
```

Comment:
given that our methods for Grouping ignore all these we can just use nonsense values...

```suggestion
    mat = StatsModels.contrasts_matrix(contrasts, baseind, 0)
```